src/i2c/src/i2c.rs

```diff
@@ -0,0 +1,548 @@
+                    2 => {
+                        data.byte = reqs[0].buf[1];
+                        I2C_SMBUS_BYTE_DATA
+                    }
+                    3 => {
+                        data.word = reqs[0].buf[1] as u16 | ((reqs[0].buf[2] as u16) << 8);
+                        I2C_SMBUS_WORD_DATA
+                    }
+                    _ => {
+                        println!(
+                            "Message length not supported for write operation: {}",
+                            reqs[0].len
+                        );
+                        return Err(Error::new(EINVAL));
+                    }
+                }
+            }
+
+            2 => {
+                if ((reqs[0].flags & I2C_M_RD) != 0)
+                    || ((reqs[1].flags & I2C_M_RD) == 0)
+                    || (reqs[0].len != 1)
+                    || (reqs[1].len > 2)
+                {
+                    println!(
+                        "Expecting a valid read smbus transfer: {:?}",
+                        (reqs.len(), reqs[0].len, reqs[1].len)
+                    );
+                    return Err(Error::new(EINVAL));
+                }
+                read_write = I2C_SMBUS_READ;
+
+                if reqs[1].len == 1 {
+                    size = I2C_SMBUS_BYTE_DATA;
+                } else {
+                    size = I2C_SMBUS_WORD_DATA;
+                }
+            }
+
+            _ => {
+                println!("Invalid number of messages for smbus xfer: {}", reqs.len());
+                return Err(Error::new(EINVAL));
+            }
+        }
+
+        Ok(I2cSmbusIoctlData {
+            read_write,
+            command: reqs[0].buf[0],
+            size,
+            data: &mut data,
+        })
+    }
+
+    /// Perform I2C_SMBUS transfer
+    fn smbus_transfer(&self, reqs: &mut [I2cReq], dryrun: bool) -> Result<()> {
+        let smbus_data = self.smbus_prepare(reqs)?;
+
+        self.do_smbus_transfer(&smbus_data, reqs[0].addr, dryrun)?;
+
+        if smbus_data.read_write == I2C_SMBUS_READ {
+            unsafe {
+                match smbus_data.size {
+                    I2C_SMBUS_BYTE => reqs[0].buf[0] = (*smbus_data.data).byte,
+                    I2C_SMBUS_BYTE_DATA => reqs[1].buf[0] = (*smbus_data.data).byte,
+                    I2C_SMBUS_WORD_DATA => {
+                        reqs[1].buf[0] = ((*smbus_data.data).word & 0xff) as u8;
+                        reqs[1].buf[1] = ((*smbus_data.data).word >> 8) as u8;
+                    }
+
+                    _ => {
+                        println!("Invalid SMBUS command: {}", smbus_data.size);
+                        return Err(Error::new(EINVAL));
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// I2C backend and helpers
+const MAX_I2C_VDEV: usize = 1 << 7;
+const I2C_INVALID_ADAPTER: u32 = 0xFFFFFFFF;
+
+pub struct I2cBackend {
+    adapters: Vec<I2cAdapter>,
```

Comment:
Seems a hashmap here may simplify the implementation:)

Reply:
> Seems a hashmap here may simplify the implementation:)

Hmm, I tried to do this but wasn't able to make it work. I tried this way (based on new patchset I just pushed, stuff got renamed):

pub struct I2cMap<A: I2cAdapterTrait> {
    adapters: HashMap<usize, &A>,
}

I had to use an reference here as I need to use same Adapter structure for multiple client devices, since they will all be accessed via the same Adapter (which has a File descriptor as well).

It then got into timeslice issues, which ones fixed got me to the fact that I am adding a local I2cAdapter variable as reference to the hashmap and got this error "returns a value referencing data owned by the current function".

I need to make sure somehow that the I2cAdapter allocated in this function (that returns the hashmap) live long enough. One way is adding them to another vector inside I2cMap struct, but that will take us to where we started from :(

All in all, I couldn't figure out how to get references stored in a hashmap to work.

@juntian I hope I have fixed all the issues you pointed out in the previous version, thanks a lot for your reviews.